[FEA] Binary IVF Flat Index

tarang-jain opened the pull request:
Depends on https://github.com/rapidsai/raft/pull/2770

Implementation of binary ivf flat index (bitwise hamming metric for the IVF Flat index)

## Key Features

### 1. **Binary Index Structure**
- Added `binary_centers_` field to store cluster centers as packed `uint8_t` arrays for binary data
- Index automatically detects BitwiseHamming metric and configures itself for binary operation
- Only support `uint8_t` inputs with BitwiseHamming and add only single instantiations of newly added kernels

### 2. **K-means Clustering for Binary Data**
The clustering approach for binary data required special handling:

- **Expanded Space Clustering**: Binary data (uint8_t) is expanded to signed representation (int8_t) where each bit becomes ±1
  - 0 → -1, 1 → +1 transformation enables meaningful centroid computation
  - Clustering performed using L2 distance in the expanded dimensional space
  
- **Centroid Quantization**: After computing float centroids in expanded space, they are converted back to binary format:
  - Centroids are stored as packed uint8_t arrays
  - KMeans (coarse) prediction is done on these quantized centroids with the BitwiseHamming distance.

### 3. **Distance Kernels**

#### Coarse Search (Cluster Selection)
- Implemented specialized `bitwise_hamming_distance_op` for query-to-centroid distances in order to compute `PairwiseDistances`

#### Fine-Grained Search (Within Clusters)  
Extended the interleaved scan kernel (`ivf_flat_interleaved_scan.cuh`) with specialized templates for BitwiseHamming:

- **Veclen-based optimization**: Different code paths based on vectorization width
  - Veclen=16,8,4: Load data as `uint32_t`, use `__popc(x ^ y)` for 4-byte Hamming distance
  - Veclen=1,2: Byte-wise XOR and population count

- **Efficient memory access patterns**: 
  - Maintains interleaved data layout for coalesced memory access
  - Specialized `loadAndComputeDist` templates for `uint8_t` that leverage vectorized loads

**as of 10/17/2025**
Binary size increase:
branch-25.12 (CUDA 12.9 + X86): 1232.414 MB
This PR (CUDA 12.9 + X86): 1251.051 MB

tarang-jain committed:
first commit

tarang-jain committed:
Merge branch 'branch-25.08' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
index header

tarang-jain committed:
populate functions;ivf_list type;kmeans_predict

tarang-jain committed:
Merge branch 'branch-25.08' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
hamming_op

tarang-jain committed:
Merge branch 'branch-25.08' into binary-kmeans

tarang-jain committed:
rm binary_ivf

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
modify ivf_flat_build

tarang-jain committed:
rm binary_ivf_flat

tarang-jain committed:
rm unused

tarang-jain committed:
updates

tarang-jain committed:
quantize

tarang-jain committed:
cleanup

tarang-jain committed:
pre-commit

tarang-jain committed:
update kmeans_predict

tarang-jain committed:
src kmeans

tarang-jain committed:
style

tarang-jain committed:
corrections to logic

tarang-jain committed:
clang

tarang-jain committed:
testing

tarang-jain committed:
correct tests

tarang-jain committed:
binary_index flag

tarang-jain committed:
correct mdspan,reduction_op

tarang-jain committed:
Merge branch 'branch-25.08' into binary-kmeans

tarang-jain committed:
correct checks

tarang-jain committed:
Merge branch 'branch-25.08' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
fused-kernel

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
cleeanup;syntax

tarang-jain committed:
cleanup

tarang-jain committed:
temp-fix-compilation-errors

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
uint32_t for storing dists

tarang-jain committed:
Merge branch 'branch-25.10' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
no use_norms check

tarang-jain committed:
coarse search

tarang-jain committed:
debug; fix dim() error

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
debug

tarang-jain committed:
Merge branch 'branch-25.10' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
fix float underflow

tarang-jain committed:
rm new file

tarang-jain committed:
Merge branch 'branch-25.10' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.10' into binary-kmeans

tarang-jain committed:
bug fixes

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' into binary-kmeans

tarang-jain committed:
correct fused instantiation

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' into binary-kmeans

tarang-jain committed:
cleanup

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'branch-25.12' into binary-kmeans

tarang-jain committed:
thorough equivalence testing checkpoint

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
cleanup-1

tarang-jain committed:
cleanup-2

tarang-jain committed:
cleanup-3

tarang-jain committed:
cleanup-4 (style

tarang-jain committed:
cleanup-6 tests

tarang-jain committed:
cleanup-7 (tests)

tarang-jain committed:
cleanup-8 tests

tarang-jain committed:
cleanup-8 (tests)

tarang-jain committed:
cleanup-8 (tests)

tarang-jain committed:
use raft ci artifacts from pr

tarang-jain committed:
wheels artifacts from raft pr

tarang-jain committed:
style

tarang-jain committed:
shell stlye check

tarang-jain committed:
Remove RAFT Dask channel from conda packages script

Removed RAFT Dask channel from the CI script.

tarang-jain committed:
fix visible sections in simt_kernel

tarang-jain committed:
Merge branch 'branch-25.12' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
raft artifact for python build

tarang-jain committed:
Merge branch 'branch-25.12' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
add skip flags

tarang-jain committed:
Merge branch 'branch-25.12' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
set binary_index_ flag explicitly

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
Merge branch 'main' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
correct recall threshold

tarang-jain committed:
merge upstream

tarang-jain committed:
style and correct tests

tarang-jain committed:
rm extra test

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
rm test

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
Merge branch 'main' into binary-kmeans

achirkin committed:
Merge branch 'main' into binary-kmeans

achirkin reviewed (changes requested):
Thanks for the thoughtful work!
I feel a little bit unsettled by the way we treat the binary distance type.
What if we use some other type to designate the binary data, such as `std::byte`  or create our own on top of `uint8_t`? Then:
  1. We could parameterise the `centers_` type to be the same type - no need for an extra member `binary_centers_`.
  2. We would not introduce more kernel instance than we already do: `index<std::byte>` would support only the Hamming distance and the other types wouldn't support that distance
  3. It would be easier to show descriptive error messages for distance type not being supported.

tarang-jain commented on cpp/src/neighbors/ivf_flat_index.cpp:
```diff
@@ -135,7 +164,11 @@ IdxT index<T, IdxT>::size() const noexcept
 {
-  return centers_.extent(1);
+  if (binary_index_) {
+    return binary_centers_.extent(1);
```
>Does that give the actual dataset dimensionality / 8?

This is correct (from a user's perspective). Essentially, number of bits / 8.

>Maybe I'm a little bit out of the context, did we have the discussion about this already?

We did not. So should I do this then:
`binary_centers_.extent(1) * 8;`

tarang-jain replied:
I have kept this as is (returns the dimension of the original dataset) and added a note in ivf_flat.hpp. Marking it as resolved.

tarang-jain commented on cpp/src/cluster/detail/kmeans_balanced.cuh:
```diff
@@ -207,6 +207,83 @@ inline std::enable_if_t<std::is_floating_point_v<MathT>> predict_core(
+                    minClusterAndDistance.data_handle(),
+                    minClusterAndDistance.data_handle() + n_rows,
+                    labels,
+                    raft::compose_op<raft::cast_op<LabelT>, raft::key_op>());
```
raft::linalg::map does not seem to work with KVP type. We are using thrust in other places too in the same file.

tarang-jain replied:
Created https://github.com/rapidsai/raft/pull/2913 to support KVP in `raft::linalg::map`

tfeher commented on cpp/src/distance/detail/distance_ops/bitwise_hamming.cuh:
```diff
@@ -0,0 +1,60 @@
+    // Ensure proper masking and casting to avoid undefined behavior
+    uint32_t xor_val    = static_cast<uint32_t>(static_cast<uint8_t>(x ^ y));
+    uint32_t masked_val = xor_val & 0xffu;
+    int popcount        = __popc(masked_val);
```
This works, but it is inefficient to use `__popc` one byte at a time. Could you create an issue to note that there is a potential improvement if we use 4 byte storage type for x and y?

tarang-jain replied:
yeah the reason I did this was to avoid separate compilation paths for larger datatypes such as half and uint32. The idea was to dispatch to the corresponding type (uint32 or half) depending on the dimension (whether it is divisible by 4 or by 2 or not). But that would bloat the binary size because the distance op would have to be defined for these additional types.

tfeher commented on cpp/src/distance/detail/fused_distance_nn.cuh:
```diff
@@ -68,16 +69,31 @@ void fusedDistanceNNImpl(OutT* min,
-      fusedCosineNN<DataT, OutT, IdxT, P, ReduceOpT, KVPReduceOpT>(
-        min, x, y, xn, yn, m, n, k, workspace, redOp, pairRedOp, sqrt, stream);
+      if constexpr (std::is_same_v<DataT, uint8_t> || std::is_same_v<DataT, int8_t>) {
+        assert(false && "Cosine distance is not supported for uint8_t/int8_t data types");
```
These `asserts` will only trigger if the code is compiled in debug mode. I don't think this is what we want.

tfeher commented on cpp/src/neighbors/ivf_flat/ivf_flat_search.cuh:
```diff
@@ -94,76 +97,113 @@ void search_impl(raft::resources const& handle,
+        distance_buffer_dev.data(),
+        uint32_distances.data(),
+        n_queries * index.n_lists(),
+        [] __device__(uint32_t val) { return static_cast<float>(val); },
```
```suggestion
        raft::cast_op<float>{},
```

tfeher reviewed (changes requested):
Thank you @tarang-jain for this PR! It looks good overall, please find a few issues below.

One thing is not clear for me: why is `uint8_t `used internally during k-means training? Each vector is a bitstring, and `uint8_t` is just a storage type that we use. We could chose `uint32_t` as well to make the popcount ops more efficient.  Since this is an implementation detail, I do not think we need to fix that in the current PR.

tarang-jain committed:
on the fly dataset expansion

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
address PR reviews

tarang-jain committed:
Merge branch 'binary-kmeans' of https://github.com/tarang-jain/cuvs into binary-kmeans

tarang-jain committed:
simplify bitwise_decode_op

tarang-jain committed:
commit suggestion

tarang-jain committed:
clang

tarang-jain commented on cpp/src/distance/detail/fused_distance_nn/fused_bitwise_hamming_nn.cuh:
```diff
@@ -0,0 +1,73 @@
+
+  dim3 blk(P::Nthreads);
+  constexpr auto maxVal = std::numeric_limits<DataT>::max();
+  typedef ::raft::KeyValuePair<IdxT, OutT> KVPair;
```
Fixed!

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
merge upstream

tarang-jain committed:
address reviews

tarang-jain committed:
style

tarang-jain commented on cpp/src/cluster/detail/kmeans_balanced.cuh:
```diff
@@ -207,6 +207,83 @@ inline std::enable_if_t<std::is_floating_point_v<MathT>> predict_core(
+}
+
+template <typename IdxT, typename LabelT>
+inline void predict_bitwise_hamming(const raft::resources& handle,
```
Because then we'd be compiling the non-BitwiseHamming paths for dtype = uint8 and distance dtype = uint32 for all the metrics. I did this for separate selective compilation.

tarang-jain commented:
@tfeher addressing your comment here:
> One thing is not clear for me: why is uint8_t used internally during k-means training? Each vector is a bitstring, and uint8_t is just a storage type that we use. We could chose uint32_t as well to make the popcount ops more efficient. Since this is an implementation detail, I do not think we need to fix that in the current PR.

The problem with that it wont allow number of bytes (dim) which is not divisible by 4 if we use uint32_t. uint8_t would be the most inclusive for dims.

tarang-jain committed:
undo ci changes

tfeher reviewed (approved):
Hi Tarang, thank you for your work on this PR, it looks good to me!

tarang-jain committed:
fix kmeans mapping op

tarang-jain committed:
other fixes to kmeans for binary data

tarang-jain committed:
fix compilation errors

tarang-jain committed:
fix compilation

tarang-jain committed:
fix compilation

tarang-jain committed:
doc

tarang-jain committed:
Merge branch 'main' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
simplify ivf-flat build

tarang-jain committed:
fix compilation errors

tarang-jain committed:
bug fixes

tarang-jain committed:
debug

tarang-jain committed:
more corrections to kmeans

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
Merge branch 'main' of https://github.com/rapidsai/cuvs into binary-kmeans

tarang-jain committed:
bug fixes

tarang-jain committed:
debug

tarang-jain committed:
working impl;rm debug statements

tarang-jain committed:
rm debug prints:

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
Merge branch 'release/26.02' into binary-kmeans

tarang-jain committed:
Merge branch 'release/26.02' into binary-kmeans

tarang-jain committed:
Merge branch 'release/26.02' into binary-kmeans

AyodeAwe committed:
REL v26.02.00 release

tarang-jain committed:
Merge branch 'release/26.02' into binary-kmeans

tarang-jain committed:
rebase

tarang-jain committed:
fix compilation

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
Merge branch 'main' into binary-kmeans

tarang-jain committed:
merge upstream, resolve merge conflicts
